dds/DCPS/WriteDataContainer.cpp

```diff
@@ -461,6 +465,10 @@ WriteDataContainer::num_all_samples()
 ACE_UINT64
 WriteDataContainer::get_unsent_data(SendStateDataSampleList& list)
 {
+  ACE_GUARD_RETURN(ACE_Recursive_Thread_Mutex,
```

Comment:
There are a few of these where the (recursive) lock is already locked on all paths into the function.

Reply:
Right, revisiting this. I didn't realize that DataWriterImpl was using the WriteDataContainer's lock instead of its own. Hooray:
1) pending_data() was always covered and never called externally, so I removed that lock and made it private. I reverted the name change since the lock is public and we can't have nice things.
2) remove_oldest_sample was also only ever called internally by dispose() (which has its own lock) and obtain_buffer() which doesn't lock but is only ever called by DataWriterImpl::write() which (ugh) locks WriteDataContainer's (public!) lock manually before calling obtain_buffer(). It's gross, but it's covered, so I removed both calls to the recursive lock in there (there was an additional one protecting a call to pending_data() in there for good measure?).
3) get_unsent_data()'s lock was secretly being made thirty levels up the stack as well, probably multiple times because why not. Removing this one as well.
4) wait_ack_of_seq(), which now touches sending_data_, needs the (non-wfa) lock and was unprotected when called by DataWriterImpl... which, by the way, looks like it's actually unlocked between the time it calls send_ack_request() and when it calls into wait_ack_of_seq() leaving room for someone else to bump the sequence number, but maybe that's not a huge deal if we're just potentially waiting on a later seq than strictly necessary? I'll defer to the existing code on this one unless I hear otherwise.